src/VersionedStates.ts

```diff
@@ -1,195 +1,140 @@
-    return this.previous.has(id)
+  has (entry: string): boolean {
+    for (const child of this) {
+      if (child === entry) return true
+    }
+    return false
   }
 
-  get (id: string): T | undefined {
-    if (id === this.id) return undefined
-    return this.previous.get(id)
+  get size (): number {
+    let size = 0
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    for (const _any of this) size += 1
+    return size
   }
 }
 
-class SetVersion <T extends string> extends DeleteVersion<T> {
-  valueSet: ReadonlySet<string>
-  state: T
+class HistoryView <State extends string> implements Version<State> {
+  timestamp: Timestamp | null
+  history: Array<Entry<State>>
 
-  constructor (timestamp: Timestamp, previous: Version<T>, id: string, state: T) {
-    super(timestamp, previous, id)
+  constructor (timestamp: Timestamp | null, history: Array<Entry<State>>) {
+    this.timestamp = timestamp
+    this.history = history
+  }
 
-    this.state = state
-    const valueSet = previous.byState(state)
-    this.valueSet = {
-      has: requestedId => {
-        if (id === requestedId) return true
-        return valueSet.has(requestedId)
-      },
-      size: valueSet.size + 1,
-      * [Symbol.iterator] () {
-        for (const requestedId of valueSet) {
-          yield requestedId
-        }
-        yield id
-      }
-    }
+  has (id: string): boolean {
+    return this.get(id) !== undefined
   }
 
-  byState (state: T): ReadonlySet<string> {
-    if (this.state === state) {
-      return this.valueSet
-    }
-    return super.byState(state)
+  byState (state: State): HistoryStateView<State> {
+    return new HistoryStateView(this, state)
   }
 
-  has (id: string): boolean {
-    if (id === this.id) return true
-    return super.has(id)
+  get (id: string): State | undefined {
+    for (const [key, value] of this) {
+      if (key === id) return value
+    }
   }
 
-  get (id: string): T | undefined {
-    if (id === this.id) return this.state
-    return super.get(id)
+  * [Symbol.iterator] (): Iterator<[key: string, value: State]> {
+    for (const [key, value] of this.iterateWithTime()) {
+      yield [key, value]
+    }
   }
 
-  * [Symbol.iterator] (): Iterator<[id: string, state: T]> {
-    const iter = super[Symbol.iterator]()
-    while (true) {
-      const entry = iter.next()
-      if (entry.done === true) break
-      else yield entry.value
+  iterateWithTime (): IterableIterator<[key: string, value: State, timestamp: Timestamp]> {
+    const alreadyFound = new Set()
+    const entries: Array<[key: string, value: State, timestamp: Timestamp]> = []
+    for (const entry of this.history) {
+      if (this.timestamp !== null && entry.timestamp.compare(this.timestamp) > 0) {
+        continue
+      }
+      if (alreadyFound.has(entry.id)) {
+        continue
+      }
+      alreadyFound.add(entry.id)
+      if (entry.state !== undefined) {
+        entries.unshift([entry.id, entry.state, entry.timestamp])
```

Comment:
Why does this need to be placed into an array instead of using `yield`?

Reply:
Sorting.

Reply:
Ah, is that so we can iterate from oldest to newest? `iterateWithTime` makes sense as a name.